### ISSUE TITLE
OSX builds extremely slow to start

### ISSUE THREAD

**Issue by awnumar** [travis-ci/travis-ci#7304]
Recently I added OSX to my [project](https://github.com/libeclipse/pocket)'s testing cycle. Previous to the change, the entire test took about 30 seconds to start, run everything, and report back. Now it takes about half an hour before even starting the build. Upon clicking the build status and going to the log, I get a message of "Hang tight, the log cannot be shown until the build has started.", while the Linux build has already completed and passed a long time ago.

Any help?

**Comment by NoahAndrews** [comment]
Look at the [status page](https://www.traviscistatus.com/#week). While the number of running Linux jobs fluctuates constantly as new VMs are created and destroyed, the number of running Mac jobs stays almost constant, which results in a big backlog. I'm sure this has to do with Mac builds having to be run on real Mac hardware. Unfortunately, this is just a thing we have to live with.

**Comment by libeclipse** [comment]
@NoahAndrews Ah right. I should hack checked that.

Closing this thread. Feel free to reopen if you feel this is a problem that needs to be addressed.

**Comment by weitjong** [comment]
I think since the slow start is a norm rather than an exception these days then perhaps travis-build should have two job queues, i.e. one queue dedicated for osx and the other one for the rest. For example, if we have 3 queued jobs for osx and 30 queued jobs for Linux; and when there is huge backlog for osx, then on the Linux side travis-build should still give us 5 concurrent build at a time to clear the 30 queued jobs. Currently it would just give us 2 (i.e. 5 - 3) concurrent build at a time. In a nutshell, our build on the Linux side is being penalized by the huge osx backlog, which is not fair at all.

**Comment by AnthonyOliveri** [comment]
For me, this is a serious issue and has a huge impact on my productivity with my Xcode projects. This has been an issue for the past few months that I've been using Travis, and has been steadily getting worse. My builds are now taking 3 hours on average to get started. 

Can someone from the Travis team explain any plans to ameliorate this situation? There is clearly a huge demand for Mac builds, with not nearly enough hardware to support it.

**Comment by libeclipse** [comment]
Keeping in mind that this is a brilliant service provided for free, the OSX build speed really is getting to the point where it's past annoying and just counterproductive. I think everyone would appreciate a comment from the Travis team about whether this is likely to be fixed, when and how.

**Comment by AnthonyOliveri** [comment]
I agree that we should be grateful that we get these builds for free, but I was told that even paid versions do nothing to speed up Mac builds. Can anyone confirm or deny that this is the case?

**Comment by scottrhoyt** [comment]
I can confirm that on a paid plan it still takes my builds about 15 - 60 minutes to start during peak times (pretty much US business hours). They start quicker off-peak but that is only marginally helpful to us. It is a becoming a big productivity drain. This is particularly true when we have many PRs that get opened simultaneously. Travis's feature of building every PR (which I love) results in non-linear growth of your build queue as you begin merging them. We have begun to be conscious of how many PRs we leave open at once for this reason, which is another hit to organization and productivity.

**Comment by loint** [comment]
OSX very slow from start. Any idea help to solve it for travis-ci.com because it's a paid service. Can not be slow by anyway !!! 👎

**Comment by loint** [comment]
@AnthonyOliveri Agree with you too.

**Comment by AnthonyOliveri** [comment]
Thank you for the clarifications Josh. The Mac builds have been much quicker over the past few days, so it looks like the infrastructure changes have helped quite a bit. I'm still getting some delays but nowhere near as bad as it had been over the past several weeks.

Is anyone else still experiencing big delays in build start times?

**Comment by matteobachetti** [comment]
Hi, 
Don't get me wrong, I love Travis and I love that it is free. So, nothing else than thanks.
However, the Mac OSX build does slow down the builds. It often starts when all other builds have finished, and it is the slowest build, so this almost doubles the build time. For the moment, I have set the OSX build as optional and I look at it by eye, this is not ideal but again... this is a free service, it's better than nothing and actually pretty awesome for Linux builds :)

**Comment by parisk** [comment]
Hello,

We are experiencing the same issue at [xterm.js](https://github.com/sourcelair/xterm.js).

An option for us would be disabling `osx` builds for all branches, but `master`. Is this something that can be configured?

I gave it a few tries with `.travis.yml`, but had no luck with it (latest [`.travis.yml`](https://github.com/sourcelair/xterm.js/blob/312e4ceb25915fae8ee6dce6dbddd19c6e6f9d7f/.travis.yml)).

**Comment by pkittenis** [comment]
Been trying to do the above at [ssh2-python](https://github.com/ParallelSSH/ssh2-python/blob/osx_ci/.travis.yml#L30-L33) but it looks like the `branches: only` block only works when set globally and is ignored in either `matrix` or `stages` `include` blocks.

**Comment by lucasbento** [comment]
Did anyone manage to do any workaround for it to get faster? I added a NodeJS project that runs on OSX, it's been 17 minutes and it didn't start yet.

**Comment by weitjong** [comment]
Look at the last two charts in https://www.traviscistatus.com/. The reason
is obvious.

Best regards,
YaoWT

**Comment by spacejam** [comment]
I'd pay like $20/mo if it meant not waiting in queues for osx builds. This is the biggest reason I'm starting to look at alternative CI systems :(

**Comment by Xaeroxe** [comment]
Hi there!  We use travis-ci.org and we greatly appreciate this awesome free service!  However given that mac builds can take in excess of 5 hours to start for us we'd like to only enable mac builds when we're concerned that a particular pull request might not work on mac.  If travis-ci gave us the ability to only use mac builds on request per PR it would greatly speed up our iteration process.  In addition to this it would likely reduce costs for travis-ci as we'd be generating significantly fewer OSX jobs.

**Comment by minggo** [comment]
The issue is not resolved. Take a look of [this build](https://travis-ci.org/cocos2d/cocos2d-x/builds/277571642?utm_source=github_status&utm_medium=notification), i restarted Mac build after waiting more than 4 hours to start, and it still being waiting.

**Comment by NoahAndrews** [comment]
@minggo They're not going to "resolve" this issue for free users. It's too expensive.

**Comment by xoviat** [comment]
One idea is for someone write an app that uses the travis-ci API to cancel OSX builds that haven't yet started if there are any queued linux builds and then restarts all of the OSX builds that it cancelled after the linux queue is processed.

It's a good idea, but not something that I really have time to do myself.

**Comment by xoviat** [comment]
Actually, on second thought, the service wouldn't need to have any state (storage). You could  just have something that looks at the builds and if linux builds are reduced from OSX usage, just cancel and restart the OSX builds, which will move them to the back of the line.

**Comment by bpoplauschi** [comment]
@joshk really appreciate your response here (and of course the help you guys give to open source projects - it does help a lot).
I want to ask if someone on your end could take a look into why there are some many builds for Mac on travis-ci.org, is it the case that there are so many projects? Or some projects just have a big number of builds?
I'm a maintainer for two big Objective-C projects ([SDWebImage](https://github.com/rs/SDWebImage) and [CocoaLumberjack](https://github.com/CocoaLumberjack/CocoaLumberjack) with thousands of stargazers) and we are somehow impacted by this. I would lie saying we can't work, but waiting for up to an hour for every build to start is a bit problematic.
Maybe there could be some priority for projects used by many people as opposed to people just trying out Travis CI and playing with it. Just a thought.
Looking forward to your opinion.

**Comment by xoviat** [comment]
I have proposed a potential solution: #8615.

@BanzaiMan Could you post a link on your website so that people can donate for that? It might be worth it if enough people donate.

**Comment by Lange** [comment]
Just throwing in my two cents: I'd gladly pay a reasonable monthly fee to get faster OSX builds for my open-source projects and help support the Travis infrastructure. I consider Travis to be a crucial bit of public infrastructure, and I'd be willing to pay a "tax" to support it just as I do any other public infrastructure.

**Comment by rraallvv** [comment]
Is there a chance to make macOS and iOS toolchains available on Linux? Maybe using something like [cctools-port](https://github.com/tpoechtrager/cctools-port).

**Comment by robertmain** [comment]
>  I'm sure this has to do with Mac builds having to be run on real Mac hardware.

wait - so does this mean that travis-ci has an iMac somewhere just running builds?

**Comment by xoviat** [comment]
> wait - so does this mean that travis-ci has an iMac somewhere just running builds?

Yes.

**Comment by robertmain** [comment]
...wow.

**Comment by Xaeroxe** [comment]
I'm fairly certain they're at least using a Mac pro.

**Comment by xoviat** [comment]
This is probably what it looks like:

![Alt text](https://www.macstadium.com/content/uploads/2016/04/Mac-Pro-Racks-wide-360w.jpg)

**Comment by robertmain** [comment]
oh ok...i guess the form factor of that is a little more managable

@xoviat  - thats cool.

**Comment by danielbayley** [comment]
I wonder if https://veertu.com/anka-technology would be relevant here?

**Comment by breznak** [comment]
Still valid pain point, have to move our (free) project from Travis, will try CircleCI

**Comment by minggo** [comment]
Yep, it is very slow. 
@breznak can CircleCI supports mac?

**Comment by rraallvv** [comment]
@minggo you might want to give Bitrise a try. It's free for macOS, and doesn't have that backlog problem.

**Comment by minggo** [comment]
Thanks @rraallvv .

**Comment by breznak** [comment]
> can CircleCI supports mac? 

@minggo for open source projects they do for free, for paid plan they of course do.

**Comment by Lyoness** [comment]
Hello Everyone,

We've (the build infra team) at Travis have been working hard to fix the performance problems of our macOS infrastructure, which meant we needed to focus all attention to migrating over to a new datacenter instead of optimizing for old hardware and SAN for better boot times.  The migration finally was implemented last week.  You'll notice that your builds no longer have wait times for private or open source projects, and build times network I/O --in addition to faster boot times-- have improved.  We're still taking in comparison data to quantify just how much of an improvement, but we've hoped you've also noticed for yourselves.  We're very happy (and relieved) to have finally shipped this.  A blog post is forthcoming. 

-Your Friendly Neighborhood Build Infrastructure Team at Travis
